pkg/tuned/controller.go

```diff
@@ -733,37 +836,49 @@ func (c *Controller) eventProcessorTuneD() {
 			return
 		}
 
-		klog.V(2).Infof("got event from workqueue")
+		klog.V(2).Infof("got event from workqueue: %#v", obj)
 		func() {
 			defer c.wqTuneD.Done(obj)
-			var workqueueKey wqKey
+			var workqueueKey wqKeyTuned
 			var ok bool
 
-			if workqueueKey, ok = obj.(wqKey); !ok {
+			if workqueueKey, ok = obj.(wqKeyTuned); !ok {
 				c.wqTuneD.Forget(obj)
-				klog.Errorf("expected wqKey in workqueue but got %#v", obj)
+				klog.Errorf("expected wqKeyTuned in workqueue but got %#v", obj)
 				return
 			}
 
+			if (c.daemon.status & scReloading) != 0 {
+				// Do not do any of the following until TuneD finished with the previous reload:
+				//   * update Profile status
+				//   * extract TuneD configuration
+				//   * reload/restart TuneD
+				// Requeuing based on TuneD reloading is far from ideal, but it is simple/reliable.
+				// When rewriting this, beware of rare deadlocks.
+				c.wqTuneD.AddRateLimited(workqueueKey)
+				return
+			}
+
+			c.change = workqueueKey.change
```

Comment:
This is not a question of _how hard_.  IIUC, you'd like to remove the ratelimiting workqueue for the `TuneD` events? If so, it could actually simplify things. However, the idea behind the ratelimiting workqueue is to "deduplicate" the same events, so that you don't need to process all of them, but just the "last" one.  This is the theory.  Now, in practise, whether it this really works and helps, hard to tell.  I've only simulated this in a contrived example. Saying that, not sure we want to make major changes to a working code that we expect to backport. But perhaps we really want to change this to make things simpler.  Happy to hear opinions.

Reply:
It is about rate limiting (process no more than N events over T time units) or about mutual exclusion? because the fact we're using a per-controller variable let me think we may have introduced a form of the latter perhaps incidentally. (thinking about https://en.wikipedia.org/wiki/Semaphore_(programming) )

Reply:
This is a sketch of what I'm suggesting. It would be really beneficial to document (with a comment) in the code why the rate limiting is needed, to shape further evolutions:
```
diff --git a/pkg/tuned/controller.go b/pkg/tuned/controller.go
index e7d89531..35fc34a9 100644
--- a/pkg/tuned/controller.go
+++ b/pkg/tuned/controller.go
@@ -133,13 +133,12 @@ type Controller struct {
 	listers *ntoclient.Listers
 	clients *ntoclient.Clients
 
-	change Change
 	daemon Daemon
 
 	tunedCmd     *exec.Cmd       // external command (tuned) being prepared or run
 	tunedExit    chan bool       // bi-directional channel to signal and register TuneD daemon exit
 	stopCh       <-chan struct{} // receive-only channel to stop the openshift-tuned controller
-	changeCh     chan bool       // bi-directional channel to wake-up the main thread to process accrued changes
+	changeCh     chan Change     // bi-directional channel to wake-up the main thread to process accrued changes
 	changeChRet  chan bool       // bi-directional channel to announce success/failure of change processing
 	tunedMainCfg *ini.File       // global TuneD configuration as defined in tuned-main.conf
 }
@@ -202,7 +201,7 @@ func newController(stopCh <-chan struct{}) (*Controller, error) {
 		clients:     clients,
 		tunedExit:   make(chan bool),
 		stopCh:      stopCh,
-		changeCh:    make(chan bool),
+		changeCh:    make(chan Change),
 		changeChRet: make(chan bool),
 	}
 
@@ -686,10 +685,10 @@ func GetBootcmdline() (string, error) {
 	return responseString, nil
 }
 
-func (c *Controller) changeSyncerProfileStatus() (synced bool) {
+func (c *Controller) changeSyncerProfileStatus(change Change) (synced bool) {
 	klog.V(2).Infof("changeSyncerProfileStatus()")
 
-	if c.change.profileStatus {
+	if change.profileStatus {
 		// One or both of the following happened:
 		// 1) tunedBootcmdlineFile changed on the filesystem.  This is very likely the result of
 		//    applying a TuneD profile by the TuneD daemon.  Make sure the node Profile k8s object
@@ -708,7 +707,7 @@ func (c *Controller) changeSyncerProfileStatus() (synced bool) {
 
 // changeSyncerTuneD synchronizes k8s objects to disk, compares them with
 // current TuneD configuration and signals TuneD process reload or restart.
-func (c *Controller) changeSyncerTuneD() (synced bool, err error) {
+func (c *Controller) changeSyncerTuneD(change Change) (synced bool, err error) {
 	var reload bool
 
 	klog.V(2).Infof("changeSyncerTuneD()")
@@ -720,20 +719,20 @@ func (c *Controller) changeSyncerTuneD() (synced bool, err error) {
 	}
 
 	// Check whether reload of the TuneD daemon is really necessary due to a Profile change.
-	if c.change.profile {
-		changeProvider, err := providerSync(c.change.provider)
+	if change.profile {
+		changeProvider, err := providerSync(change.provider)
 		if err != nil {
 			return false, err
 		}
 		reload = reload || changeProvider
 
-		if c.daemon.recommendedProfile != c.change.recommendedProfile {
-			if err = TunedRecommendFileWrite(c.change.recommendedProfile); err != nil {
+		if c.daemon.recommendedProfile != change.recommendedProfile {
+			if err = TunedRecommendFileWrite(change.recommendedProfile); err != nil {
 				return false, err
 			}
-			klog.Infof("recommended TuneD profile changed from (%s) to (%s)", c.daemon.recommendedProfile, c.change.recommendedProfile)
+			klog.Infof("recommended TuneD profile changed from (%s) to (%s)", c.daemon.recommendedProfile, change.recommendedProfile)
 			// Cache the value written to tunedRecommendFile.
-			c.daemon.recommendedProfile = c.change.recommendedProfile
+			c.daemon.recommendedProfile = change.recommendedProfile
 			reload = true
 		} else {
 			klog.Infof("recommended profile (%s) matches current configuration", c.daemon.recommendedProfile)
@@ -747,10 +746,10 @@ func (c *Controller) changeSyncerTuneD() (synced bool, err error) {
 
 		// Does the current TuneD process have debugging turned on?
 		debug := (c.daemon.restart & ctrlDebug) != 0
-		if debug != c.change.debug {
+		if debug != change.debug {
 			// A complete restart of the TuneD daemon is needed due to a debugging request switched on or off.
 			c.daemon.restart |= ctrlRestart
-			if c.change.debug {
+			if change.debug {
 				c.daemon.restart |= ctrlDebug
 			} else {
 				c.daemon.restart &= ^ctrlDebug
@@ -759,7 +758,7 @@ func (c *Controller) changeSyncerTuneD() (synced bool, err error) {
 
 		// Does the current TuneD process have the reapply_sysctl option turned on?
 		reapplySysctl := c.tunedMainCfg.Section("").Key("reapply_sysctl").MustBool()
-		if reapplySysctl != c.change.reapplySysctl {
+		if reapplySysctl != change.reapplySysctl {
 			if err = iniCfgSetKey(c.tunedMainCfg, "reapply_sysctl", !reapplySysctl); err != nil {
 				return false, err
 			}
@@ -771,7 +770,7 @@ func (c *Controller) changeSyncerTuneD() (synced bool, err error) {
 		}
 	}
 
-	if c.change.rendered {
+	if change.rendered {
 		// The "rendered" Tuned k8s object changed.
 		tuned, err := c.listers.TunedResources.Get(tunedv1.TunedRenderedResourceName)
 		if err != nil {
@@ -790,7 +789,7 @@ func (c *Controller) changeSyncerTuneD() (synced bool, err error) {
 	if reload {
 		c.daemon.restart |= ctrlReload
 	}
-	err = c.changeSyncerRestartOrReloadTuneD()
+	err = changeSyncerRestartOrReloadTuneD()
 
 	return err == nil, err
 }
@@ -817,14 +816,14 @@ func (c *Controller) changeSyncerRestartOrReloadTuneD() error {
 // reloads as needed.  Returns indication whether the change was successfully
 // synced and an error.  Only critical errors are returned, as non-nil errors
 // will cause restart of the main control loop -- the changeWatcher() method.
-func (c *Controller) changeSyncer() (synced bool, err error) {
+func (c *Controller) changeSyncer(change Change) (synced bool, err error) {
 	// Sync k8s Profile status if/when needed.
-	if !c.changeSyncerProfileStatus() {
+	if !c.changeSyncerProfileStatus(change) {
 		return false, nil
 	}
 
 	// Extract TuneD configuration from k8s objects and restart/reload TuneD as needed.
-	return c.changeSyncerTuneD()
+	return c.changeSyncerTuneD(change)
 }
 
 // eventProcessorTuneD is a long-running method that will continually
@@ -860,8 +859,7 @@ func (c *Controller) eventProcessorTuneD() {
 				return
 			}
 
-			c.change = workqueueKey.change
-			c.changeCh <- true
+			c.changeCh <- workqueueKey.change
 			eventProcessed := <-c.changeChRet
 			if !eventProcessed {
 				requeued := c.wqTuneD.NumRequeues(workqueueKey)
@@ -1147,11 +1145,11 @@ func (c *Controller) changeWatcher() (err error) {
 		case err := <-wFs.Errors:
 			return fmt.Errorf("error watching filesystem: %v", err)
 
-		case <-c.changeCh:
+		case ch := <-c.changeCh:
 			var synced bool
 			klog.V(2).Infof("changeCh")
 
-			synced, err := c.changeSyncer()
+			synced, err := c.changeSyncer(ch)
 			if err != nil {
 				return err
 			}
```
the reason why I'm suggesting this change is to go in the direction of reducing the common/implicit state, because doing so very very often leads to easier and simpler code.

Reply:
> It is about rate limiting (process no more than N events over T time units) or about mutual exclusion? 

We can probably forget about rate limiting in wqTuneD workqueue.  That's just a convenience that will rarely happen in my view.  I'd say rate limiting only needs to be done for k8s object changes in wqKube workqueue.  What needs to be handled are:
  * requeues of failed attempts at updating profile status for the operator
  * mutual exclusion (not to extract any TuneD config during its reloads/restart and not to reload/restart during ongoing reloads/restarts)

Reply:
> This is a sketch of what I'm suggesting. Utterly untested. It would be really beneficial to document (with a comment) in the code why the rate limiting is needed, to shape further evolutions:
> 
> CODE CUT FOR BREVITY
> the reason why I'm suggesting this change is to go in the direction of reducing the common/implicit state, because doing so very very often leads to easier and simpler code.

Oh, I see what you mean now, I was being a bit dumb.  I like the change and thank you for it.  Do you want to co-author or shall I just adapt it?  Testing the code now.

Reply:
no need to co-author me for such a simple change. Feel free to take the code and let's move on.